File: sample-tests/src/test-generator.ts
import { basename, dirname } from 'path'
import type { ExtractedBlock } from './mdx-extractor'
import { stripImportsFromBlock } from './mdx-extractor'

export interface GenerateTestFileOptions {
  sourcePath: string
  blocks: ExtractedBlock[]
  importSuperset: string
  helperModulePath?: string
  /** Map from test name → list of variable names the pre-test provides */
  preTestProvides?: Record<string, string[]>
}

function computeSdkFixtureRelativePath(sourcePath: string): string {
  const withoutSdk = sourcePath.replace(/^sdk\//, '')
  const dir = dirname(withoutSdk)
  const depth = dir === '.' ? 0 : dir.split('/').length
  const ups = '../'.repeat(depth + 1)
  return `${ups}helpers/sdk-fixture`
}

/**
 * Extract top-level function declaration names from a code block.
 * Matches patterns like:
 *   async function foo(...)
 *   function bar(...)
 */
export function extractFunctionNames(code: string): string[] {
  const names: string[] = []
  const re = /^(?:async\s+)?function\s+(\w+)\s*\(/gm
  let m: RegExpExecArray | null
  while ((m = re.exec(code)) !== null) {
    names.push(m[1])
  }
  return names
}

/**
 * Extract top-level const/let variable declaration names from a code block.
 * Matches patterns like:
 *   const FOO = ...
 *   let bar = ...
 */
export function extractVariableNames(code: string): string[] {
  const names: string[] = []
  const re = /^(?:const|let)\s+(\w+)\s*=/gm
  let m: RegExpExecArray | null
  while ((m = re.exec(code)) !== null) {
    names.push(m[1])
  }
  return names
}

export function generateTestFileContent(options: GenerateTestFileOptions): string {
  const { sourcePath, blocks, importSuperset, helperModulePath, preTestProvides } = options
  const fileBaseName = basename(sourcePath, '.mdx')
  const sdkFixturePath = computeSdkFixtureRelativePath(sourcePath)

  const lines: string[] = []

  lines.push(`// Auto-generated from ${sourcePath} — do not edit`)
  lines.push(importSuperset)
  lines.push(`import { getTestSdk } from "${sdkFixturePath}"`)
  if (helperModulePath !== undefined) {
    lines.push(`import { preTest, postTest } from "${helperModulePath}"`)
  }
  lines.push('')
  lines.push('let sdk: CardinalSdk')
  lines.push('')
  lines.push('beforeAll(async () => {')
  lines.push('  sdk = await getTestSdk()')
  lines.push('})')
  lines.push('')

  for (let i = 0; i < blocks.length; i++) {
    const block = blocks[i]
    const testName = `${fileBaseName} block ${i + 1} (line ${block.startLine})`
    const body = stripImportsFromBlock(block.code)
    const indentedBody = body
      .split('\n')
      .map(line => (line.trim() === '' ? '' : `  ${line}`))
      .join('\n')

    lines.push(`test("${testName}", async () => {`)

    if (helperModulePath !== undefined) {
      const provides = preTestProvides?.[testName] ?? []
      if (provides.length > 0) {
        lines.push(`  const { ${provides.join(', ')} } = await preTest["${testName}"]?.(sdk) ?? {}`)
      } else {
        lines.push(`  await preTest["${testName}"]?.(sdk)`)
      }
    }

    lines.push(indentedBody)

    if (helperModulePath !== undefined) {
      const funcNames = extractFunctionNames(body)
      const varNames = extractVariableNames(body)
      const allNames = [...funcNames, ...varNames]
      const args = ['sdk', ...allNames].join(', ')
      lines.push(`  await postTest["${testName}"]?.(${args})`)
    }

    lines.push('})')
    lines.push('')
  }

  return lines.join('\n')
}

File: sample-tests/kotlin/src/main/kotlin/com/icure/docs/generator/TestGenerator.kt
package com.icure.docs.generator

import java.io.File

data class GenerateTestFileOptions(
    val sourcePath: String,
    val blocks: List<ExtractedBlock>,
    val importSuperset: String,
    val helperClassName: String?,
    val helperImportPath: String?,
    val preTestProvides: Map<String, List<String>>,
    val preTestProvideTypes: Map<String, Map<String, String>> = emptyMap(),
)

private val FUNCTION_REGEX = Regex("""^(?:suspend\s+)?fun\s+(\w+)\s*\(""", RegexOption.MULTILINE)
private val VARIABLE_REGEX = Regex("""^(?:val|var)\s+(\w+)\s*[=:]""", RegexOption.MULTILINE)

fun extractFunctionNames(code: String): List<String> {
    return FUNCTION_REGEX.findAll(code).map { it.groupValues[1] }.toList()
}

fun extractVariableNames(code: String): List<String> {
    return VARIABLE_REGEX.findAll(code).map { it.groupValues[1] }.toList()
}

fun sourcePathToPackage(sourcePath: String): String {
    val withoutSdk = sourcePath.removePrefix("sdk/")
    val dir = File(withoutSdk).parent ?: return "com.icure.docs.generated"
    val packageSuffix = dir
        .replace(File.separatorChar, '.')
        .replace('/', '.')
        .replace("-", "")
        .lowercase()
    return "com.icure.docs.generated.$packageSuffix"
}

fun sourcePathToClassName(sourcePath: String): String {
    val baseName = File(sourcePath).nameWithoutExtension
    val raw = baseName
        .split(Regex("[_\\-]"))
        .joinToString("") { segment ->
            segment.replaceFirstChar { it.uppercaseChar() }
        } + "Test"
    // Kotlin class names cannot start with a digit
    return if (raw.first().isDigit()) "_$raw" else raw
}

fun fileBaseNameFromSource(sourcePath: String): String {
    return File(sourcePath).nameWithoutExtension
}

fun generateTestFileContent(options: GenerateTestFileOptions): String {
    val (sourcePath, blocks, importSuperset, helperClassName, helperImportPath, preTestProvides, preTestProvideTypes) = options
    val fileBaseName = fileBaseNameFromSource(sourcePath)
    val packageName = sourcePathToPackage(sourcePath)
    val className = sourcePathToClassName(sourcePath)

    val lines = mutableListOf<String>()

    lines.add("// Auto-generated from $sourcePath -- do not edit")
    lines.add("@file:OptIn(kotlin.time.ExperimentalTime::class)")
    lines.add("@file:Suppress(\"LocalVariableName\", \"unused\", \"UNUSED_VARIABLE\", \"RemoveRedundantQualifierName\", \"UNCHECKED_CAST\", \"UNREACHABLE_CODE\", \"RedundantSuspendModifier\")")
    lines.add("package $packageName")
    lines.add("")
    lines.add(importSuperset)
    lines.add("")
    lines.add("import com.icure.docs.helpers.SdkFixture")
    if (helperImportPath != null) {
        lines.add("import $helperImportPath")
    }
    lines.add("import io.kotest.core.spec.style.FunSpec")
    lines.add("")
    lines.add("class $className : FunSpec({")
    lines.add("")
    lines.add("    lateinit var sdk: com.icure.cardinal.sdk.CardinalSdk")
    lines.add("")
    lines.add("    beforeSpec {")
    lines.add("        sdk = SdkFixture.getTestSdk()")
    lines.add("    }")
    lines.add("")

    for ((i, block) in blocks.withIndex()) {
        val blockRef = block.testId ?: "line ${block.startLine}"
        val testName = "$fileBaseName block ${i + 1} ($blockRef)"
        val body = stripImportsFromBlock(block.code)
        val indentedBody = body.split("\n").joinToString("\n") { line ->
            if (line.isBlank()) "" else "        $line"
        }

        lines.add("    test(\"$testName\") {")

        if (helperClassName != null) {
            val provides = preTestProvides[testName] ?: emptyList()
            if (provides.isNotEmpty()) {
                lines.add("        val __provided = ${helperClassName}.preTest[\"$testName\"]?.invoke(sdk) ?: emptyMap()")
                val typeMap = preTestProvideTypes[testName] ?: emptyMap()
                for (varName in provides) {
                    val type = typeMap[varName]
                    if (type != null) {
                        lines.add("        val $varName = __provided[\"$varName\"] as $type")
                    } else {
                        lines.add("        val $varName = __provided[\"$varName\"]")
                    }
                }
            } else {
                lines.add("        ${helperClassName}.preTest[\"$testName\"]?.invoke(sdk)")
            }
        }

        lines.add(indentedBody)

        if (helperClassName != null) {
            val funcNames = extractFunctionNames(body)
            val varNames = extractVariableNames(body)
            val allNames = funcNames + varNames
            if (allNames.isNotEmpty()) {
                lines.add("        val __extracted = mutableMapOf<String, Any?>()")
                for (name in funcNames) {
                    lines.add("        __extracted[\"$name\"] = ::$name")
                }
                for (name in varNames) {
                    lines.add("        __extracted[\"$name\"] = $name")
                }
                lines.add("        ${helperClassName}.postTest[\"$testName\"]?.invoke(sdk, __extracted)")
            } else {
                lines.add("        ${helperClassName}.postTest[\"$testName\"]?.invoke(sdk, emptyMap())")
            }
        }

        lines.add("    }")
        lines.add("")
    }

    lines.add("})")
    lines.add("")

    return lines.joinToString("\n")
}
